uaclient/tests/test_gpg.py

```diff
@@ -0,0 +1,70 @@
+from uaclient import exceptions, gpg, util
+from uaclient.testing import data
+
+import logging
+import os
+import pytest
+
+
+class TestExportGPGKeyFromKeyring:
+
+    def test_key_error_on_missing_key(self, tmpdir):
+        """Raise UserFacingError when key_id is not in source_keyring_file."""
+        src_keyring = tmpdir.join('ubuntu-advantage-keyring')
+        src_keyring.write('')
+        destination_keyfile = tmpdir.join('destination_keyfile').strpath
+        key_id = 'Bueller... Bueller'
+        with pytest.raises(exceptions.UserFacingError) as excinfo:
+            gpg.export_gpg_key_from_keyring(
+                key_id=key_id, source_keyring_file=src_keyring.strpath,
+                destination_keyfile=destination_keyfile)
+
+        error_msg = "GPG key '%s' not found in %s" % (
+            key_id, src_keyring.strpath)
+        assert error_msg in str(excinfo.value)
+        assert False is os.path.exists(destination_keyfile)
+
+    @pytest.mark.parametrize('caplog_text', [logging.ERROR], indirect=True)
+    def test_user_facing_error_on_invalid_keyring_packet(
+            self, caplog_text, tmpdir):
+        """Raise UserFacingError on invalid keyring packet. Log cmd error."""
+        source_keyring = tmpdir.join('keyring').strpath
+        destination_keyfile = tmpdir.join('destination_key').strpath
+        key_id = '3CB3DF682220A643B43065E9B30EDAA63D8F61D0'
+        # Create bogus keyring with invalid packet which causes gpg errors
+        with open(source_keyring, 'wb') as stream:
+            stream.write(b'\n')
+        with pytest.raises(exceptions.UserFacingError) as excinfo:
+            gpg.export_gpg_key_from_keyring(
+                key_id=key_id, source_keyring_file=source_keyring,
+                destination_keyfile=destination_keyfile)
+        msg = "Failed running command 'gpg --output {}".format(
+            destination_keyfile)
+        assert 'ERROR    {}'.format(msg) in caplog_text()
+        assert False is os.path.exists(destination_keyfile)
+        msg = 'Unable to export GPG keys from keyring %s' % source_keyring
```

Comment:
https://write.wrestle.town/oddbloke/python-anti-pattern-single-element-s-formatting

Reply:
thanks for the blog link, now I get to have some fun reading your blog.
fixed.